### PR TITLE
connector: implement LinkedIn connector

### PR DESCRIPTION
**Pull request by pborzenkov**
connector/linkedin implements authorization strategy via LinkedIn's
OAuth2 endpoint + profile API.

It doesn't implement RefreshConnector as LinkedIn doesn't provide any
refresh token at all (https://developer.linkedin.com/docs/oauth2, Step 5
— Refresh your Access Tokens) and recommends ordinary AuthCode exchange
flow when token refresh is required.

### CODE REVIEW

File: connector/linkedin/linkedin.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	ctx := r.Context()
+	token, err := c.oauth2Config.Exchange(ctx, q.Get("code"))
+	if err != nil {
+		return identity, fmt.Errorf("linkedin: get token: %v", err)
+	}
+
+	client := c.oauth2Config.Client(ctx, token)
+	profile, err := c.profile(ctx, client)
+	if err != nil {
+		return identity, fmt.Errorf("github: get profile: %v", err)
```
Comment (ericchiang) [inline connector/linkedin/linkedin.go]: nit: "github"

File: connector/linkedin/linkedin.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	identity = connector.Identity{
+		UserID:        profile.ID,
+		Username:      profile.FirstName + " " + profile.LastName,
```
Comment (ericchiang) [inline connector/linkedin/linkedin.go]: nit: we should be smarter about this value. e.g. ensure that the FirstName and LastName are actually set.

File: connector/linkedin/linkedin.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		Email:         profile.Email,
```
Comment (ericchiang) [inline connector/linkedin/linkedin.go]: Please error if profile.Email isn't set.

File: connector/linkedin/linkedin.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+type profile struct {
+	ID        string `json:"id"`
+	FirstName string `json:"firstName"`
+	LastName  string `json:"lastName"`
+	Email     string `json:"emailAddress"`
+}
+
+func (c *linkedInConnector) profile(ctx context.Context, client *http.Client) (p profile, err error) {
+	// https://developer.linkedin.com/docs/fields/basic-profile
+	req, err := http.NewRequest("GET", apiURL+"/people/~:(id,first-name,last-name,email-address)", nil)
```
Comment (ericchiang) [inline connector/linkedin/linkedin.go]: this is a really weird URL. ha

File: connector/linkedin/linkedin.go (hunk 3 above)
Comment (pborzenkov) [reply connector/linkedin/linkedin.go]: That's seems like an impossible condition. This is the main email that was used for registration and we always ask permission for r_emailaddress claim.
I'd add the check just to be sure anyway.

File: connector/linkedin/linkedin.go (hunk 2 above)
Comment (pborzenkov) [reply connector/linkedin/linkedin.go]: ok

File: connector/linkedin/linkedin.go (hunk 1 above)
Comment (pborzenkov) [reply connector/linkedin/linkedin.go]: Will fix.

**Comment by ericchiang** [comment]
This conflicts with some additions in https://github.com/coreos/dex/pull/1100

Mind rebasing?

**Pull request merged**